src/backend/logger/logger.ts

```diff
@@ -355,10 +356,94 @@ export function logFileLocation(appName: string) {
   return join(gamesConfigPath, `${appName}.log`)
 }
 
+const logsWriters: Record<string, LogWriter> = {}
+
+class LogWriter {
+  appName: string
+  queue: string[] | undefined
+  initialized: boolean
+  timeoutId: NodeJS.Timeout | undefined
+  filePath: string
+
+  constructor(appName: string) {
+    this.appName = appName
+    this.initialized = false
+    this.filePath = lastPlayLogFileLocation(appName)
+  }
+
+  logMessage(message: string) {
+    this.queue ??= []
+
+    // push messages to append to the log
+    this.queue.push(message)
+
+    // if the logger is initialized and we don't have a timeout,
+    // append the message and start a timeout
+    //
+    // otherwise it means there's a timeout already running that will
+    // write the elements in the queue in a second
+    if (this.initialized && !this.timeoutId) this.appendMessages()
+  }
+
+  async initLog() {
+    try {
+      const info = await getSystemInfo()
+      const systemInfo = await formatSystemInfo(info)
+
+      // init log file and then append message if any
+      try {
+        await writeFile(
+          this.filePath,
+          'System Info:\n' + `${systemInfo}\n` + '\n'
+        )
+
+        this.initialized = true
+        this.appendMessages()
+      } catch (error) {
+        logError(
+          [`Failed to initialize log ${this.filePath}:`, error],
+          LogPrefix.Backend
+        )
+      }
+    } catch (error) {
+      logError(['Failed to fetch system information', error], LogPrefix.Backend)
+    }
```

Comment:
The try-catch blocks here are now a little hard to follow

```suggestion
    let systemInfo
    try {
      systemInfo = await getSystemInfo().then(formatSystemInfo)
    } catch (error) {
      logError(['Failed to fetch system information', error], LogPrefix.Backend)
      return
    }

    // init log file and then append message if any
    try {
      await writeFile(
        this.filePath,
        'System Info:\n' + `${systemInfo}\n` + '\n'
      )

      this.initialized = true
      this.appendMessages()
    } catch (error) {
      logError(
        [`Failed to initialize log ${this.filePath}:`, error],
        LogPrefix.Backend
      )
    }
```

Reply:
I'm refactoring that in https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/3394/files#diff-3b36a7de185377bbb618ac5e7bc20b0780c6ea78b5d5ac53c7b9a7469448a988L388, what do you think?

(also using `writeFileSync`, realized there's no need to do `await writeFile` if we can just use the sync function?)

Reply:
That's much more readable IMO, thanks

*As far as I know* `writeFileSync` blocks execution until the write operation is finished, meanwhile Promise-based APIs are "transparent" to the JS engine (so it knows that we're waiting on something and can in the meantime handle other things, like callbacks or IPC events)